drivers/adc/adc_stm32.c

```diff
@@ -823,6 +867,79 @@ static int set_resolution(const struct device *dev,
 	return 0;
 }
 
+static int set_sequencer(const struct device *dev)
+{
+	const struct adc_stm32_cfg *config = dev->config;
+	struct adc_stm32_data *data = dev->data;
+	ADC_TypeDef *adc = (ADC_TypeDef *)config->base;
+
+	uint8_t channel_id;
+	uint8_t channel_index = 0;
+	uint32_t channels_mask = 0;
+
+	/* Iterate over selected channels in bitmask keeping track of:
+	 * - channel_index: ranging from 0 -> ( data->channel_count - 1 )
+	 * - channel_id: ordinal position of channel in data->channels bitmask
+	 */
+	for (uint32_t channels = data->channels; channels;
+		      channels &= ~BIT(channel_id), channel_index++) {
+		channel_id = find_lsb_set(channels) - 1;
+
+		uint32_t channel = __LL_ADC_DECIMAL_NB_TO_CHANNEL(channel_id);
+		channels_mask |= channel;
+
+		adc_stm32_setup_channel(dev, channel_id);
+
+#if defined(FULLY_CONFIGURABLE_SEQUENCER)
+#if defined(CONFIG_SOC_SERIES_STM32U5X)
+		/* On STM32U5 ADC4 is not fully configurable, and has no preselection */
+		if (adc != ADC4)
+#endif /* CONFIG_SOC_SERIES_STM32U5X */
+		{
+#if defined(CONFIG_SOC_SERIES_STM32H7X) || defined(CONFIG_SOC_SERIES_STM32U5X)
+			/*
+			* Each channel in the sequence must be previously enabled in PCSEL.
+			* This register controls the analog switch integrated in the IO level.
+			*/
+			LL_ADC_SetChannelPreselection(adc, channel);
+#endif /* CONFIG_SOC_SERIES_STM32H7X || CONFIG_SOC_SERIES_STM32U5X */
+			LL_ADC_REG_SetSequencerRanks(adc, table_rank[channel_index], channel);
+			LL_ADC_REG_SetSequencerLength(adc, table_seq_len[channel_index]);
+		}
+#endif /* FULLY_CONFIGURABLE_SEQUENCER */
+	}
+
+#if defined(NOT_FULLY_CONFIGURABLE_SEQUENCER)
+#if defined(CONFIG_SOC_SERIES_STM32U5X)
+	/* On STM32U5 only ADC4 is not fully configurable */
+	if (adc == ADC4)
+#endif /* CONFIG_SOC_SERIES_STM32U5X */
```

Comment:
Looks like defining this in dt would simplify code a bit

Reply:
I could add a property for the type of sequencer, but some functions don't exist in some series, so I still need an #ifdef at some point.

Reply:
Though I suppose I could create a macro similar to ANY_NUM_COMMON_SAMPLING_TIME_CHANNELS_IS to check if any ADC are either fully configurable or not to exclude the functions not present for the series.
I'll have a look at what is possible